File: scanner.php
<?php
header('Content-Type: application/json');
header('Access-Control-Allow-Origin: *');
header('Access-Control-Allow-Methods: POST, GET, OPTIONS');
header('Access-Control-Allow-Headers: Content-Type');

if ($_SERVER['REQUEST_METHOD'] === 'OPTIONS') {
    exit(0);
}

class WebsiteScanner {
    private $url;
    private $domain;
    private $results = [];

    public function __construct($url) {
        $this->url = $url;
        $this->domain = parse_url($url, PHP_URL_HOST);
    }

    public function scan() {
        try {
            $this->results['basic_info'] = $this->getBasicInfo();
            $this->results['server_info'] = $this->getServerInfo();
            $this->results['domain_info'] = $this->getDomainInfo();
            $this->results['security_info'] = $this->getSecurityInfo();
            $this->results['performance_info'] = $this->getPerformanceInfo();
            $this->results['technology_info'] = $this->getTechnologyInfo();
            
            return [
                'success' => true,
                'data' => $this->results,
                'scan_time' => date('Y-m-d H:i:s')
            ];
        } catch (Exception $e) {
            return [
                'success' => false,
                'error' => $e->getMessage()
            ];
        }
    }

    private function getBasicInfo() {
        $headers = @get_headers($this->url, 1);
        $status_code = $headers ? $this->extractStatusCode($headers[0]) : 'Unknown';
        
        return [
            'url' => $this->url,
            'domain' => $this->domain,
            'status_code' => $status_code,
            'protocol' => parse_url($this->url, PHP_URL_SCHEME),
            'port' => parse_url($this->url, PHP_URL_PORT) ?: (parse_url($this->url, PHP_URL_SCHEME) === 'https' ? 443 : 80)
        ];
    }

    private function getServerInfo() {
        $ip = gethostbyname($this->domain);
        $headers = @get_headers($this->url, 1);
        
        $server = 'Unknown';
        if ($headers && isset($headers['Server'])) {
            $server = is_array($headers['Server']) ? $headers['Server'][0] : $headers['Server'];
        }

        // Get geolocation info (simplified)
        $location_info = $this->getLocationInfo($ip);
        
        return [
            'ip_address' => $ip,
            'server' => $server,
            'location' => $location_info['country'] ?? 'Unknown',
            'city' => $location_info['city'] ?? 'Unknown',
            'isp' => $location_info['isp'] ?? 'Unknown'
        ];
    }

    private function getDomainInfo() {
        // Simplified domain info (in real implementation, you'd use WHOIS API)
        return [
            'registrar' => 'Unknown',
            'creation_date' => 'Unknown',
            'expiry_date' => 'Unknown',
            'name_servers' => $this->getNameServers(),
            'dnssec' => 'Unknown'
        ];
    }

    private function getSecurityInfo() {
        $ssl_info = $this->getSSLInfo();
        $security_headers = $this->getSecurityHeaders();
        
        return [
            'ssl_certificate' => $ssl_info,
            'security_headers' => $security_headers,
            'hsts_enabled' => isset($security_headers['Strict-Transport-Security']),
            'content_security_policy' => isset($security_headers['Content-Security-Policy'])
        ];
    }

    private function getPerformanceInfo() {
        $start_time = microtime(true);
        $headers = @get_headers($this->url, 1);
        $response_time = (microtime(true) - $start_time) * 1000;
        
        return [
            'response_time' => round($response_time, 2) . 'ms',
            'compression' => $this->checkCompression($headers),
            'caching' => $this->checkCaching($headers),
            'page_size' => $this->getPageSize()
        ];
    }

    private function getTechnologyInfo() {
        $headers = @get_headers($this->url, 1);
        $content = @file_get_contents($this->url);
        
        return [
            'web_server' => $this->detectWebServer($headers),
            'programming_language' => $this->detectProgrammingLanguage($headers, $content),
            'cms' => $this->detectCMS($content),
            'javascript_frameworks' => $this->detectJSFrameworks($content),
            'analytics' => $this->detectAnalytics($content)
        ];
    }

    private function extractStatusCode($header) {
        preg_match('/HTTP\/\d\.\d\s+(\d+)/', $header, $matches);
        return isset($matches[1]) ? $matches[1] : 'Unknown';
    }

    private function getLocationInfo($ip) {
        // Simplified location detection
        // In real implementation, use services like ipapi.co or similar
        return [
            'country' => 'Unknown',
            'city' => 'Unknown',
            'isp' => 'Unknown'
        ];
    }

    private function getNameServers() {
        $nameservers = [];
        $dns_records = @dns_get_record($this->domain, DNS_NS);
        
        if ($dns_records) {
            foreach ($dns_records as $record) {
                $nameservers[] = $record['target'];
            }
        }
        
        return implode(', ', $nameservers) ?: 'Unknown';
    }

    private function getSSLInfo() {
        if (parse_url($this->url, PHP_URL_SCHEME) !== 'https') {
            return ['enabled' => false, 'message' => 'SSL not enabled'];
        }

        $context = stream_context_create([
            'ssl' => [
                'capture_peer_cert' => true,
                'verify_peer' => false,
                'verify_peer_name' => false
            ]
        ]);

        $stream = @stream_socket_client(
            'ssl://' . $this->domain . ':443',
            $errno, $errstr, 30, STREAM_CLIENT_CONNECT, $context
        );

        if (!$stream) {
            return ['enabled' => false, 'message' => 'SSL connection failed'];
        }

        $cert = stream_context_get_params($stream)['options']['ssl']['peer_certificate'];
        $cert_info = openssl_x509_parse($cert);

        fclose($stream);

        return [
            'enabled' => true,
            'issuer' => $cert_info['issuer']['CN'] ?? 'Unknown',
            'valid_from' => date('Y-m-d', $cert_info['validFrom_time_t']),
            'valid_to' => date('Y-m-d', $cert_info['validTo_time_t']),
            'subject' => $cert_info['subject']['CN'] ?? 'Unknown'
        ];
    }

    private function getSecurityHeaders() {
        $headers = @get_headers($this->url, 1);
        $security_headers = [];
        
        $security_header_names = [
            'Strict-Transport-Security',
            'Content-Security-Policy',
            'X-Frame-Options',
            'X-Content-Type-Options',
            'X-XSS-Protection',
            'Referrer-Policy'
        ];

        foreach ($security_header_names as $header_name) {
            if (isset($headers[$header_name])) {
                $security_headers[$header_name] = is_array($headers[$header_name]) 
                    ? $headers[$header_name][0] 
                    : $headers[$header_name];
            }
        }

        return $security_headers;
    }

    private function checkCompression($headers) {
        return isset($headers['Content-Encoding']) ? 'Enabled' : 'Disabled';
    }

    private function checkCaching($headers) {
        $cache_headers = ['Cache-Control', 'Expires', 'ETag', 'Last-Modified'];
        
        foreach ($cache_headers as $header) {
            if (isset($headers[$header])) {
                return 'Enabled';
            }
        }
        
        return 'Disabled';
    }

    private function getPageSize() {
        $content = @file_get_contents($this->url);
        return $content ? round(strlen($content) / 1024, 2) . ' KB' : 'Unknown';
    }

    private function detectWebServer($headers) {
        if (isset($headers['Server'])) {
            $server = is_array($headers['Server']) ? $headers['Server'][0] : $headers['Server'];
            return explode('/', $server)[0];
        }
        return 'Unknown';
    }

    private function detectProgrammingLanguage($headers, $content) {
        // Check headers for language indicators
        if (isset($headers['X-Powered-By'])) {
            $powered_by = is_array($headers['X-Powered-By']) ? $headers['X-Powered-By'][0] : $headers['X-Powered-By'];
            if (stripos($powered_by, 'PHP') !== false) return 'PHP';
            if (stripos($powered_by, 'ASP.NET') !== false) return 'ASP.NET';
        }

        // Check content for language indicators
        if ($content) {
            if (preg_match('/\.php["\']/', $content)) return 'PHP';
            if (preg_match('/\.aspx["\']/', $content)) return 'ASP.NET';
            if (preg_match('/\.jsp["\']/', $content)) return 'Java';
            if (preg_match('/\.py["\']/', $content)) return 'Python';
        }

        return 'Unknown';
    }

    private function detectCMS($content) {
        if (!$content) return 'Unknown';

        $cms_patterns = [
            'WordPress' => '/wp-content|wp-includes|wordpress/i',
            'Drupal' => '/drupal|sites\/default/i',
            'Joomla' => '/joomla|option=com_/i',
            'Magento' => '/magento|mage\/js/i',
            'Shopify' => '/shopify|cdn\.shopify/i'
        ];

        foreach ($cms_patterns as $cms => $pattern) {
            if (preg_match($pattern, $content)) {
                return $cms;
            }
        }

        return 'Custom/Unknown';
    }

    private function detectJSFrameworks($content) {
        if (!$content) return 'Unknown';

        $frameworks = [];
        $framework_patterns = [
            'React' => '/react|reactjs/i',
            'Vue.js' => '/vue\.js|vuejs/i',
            'Angular' => '/angular|ng-/i',
            'jQuery' => '/jquery/i'
        ];

        foreach ($framework_patterns as $framework => $pattern) {
            if (preg_match($pattern, $content)) {
                $frameworks[] = $framework;
            }
        }

        return empty($frameworks) ? 'Unknown' : implode(', ', $frameworks);
    }

    private function detectAnalytics($content) {
        if (!$content) return 'Unknown';

        $analytics_patterns = [
            'Google Analytics' => '/google-analytics|gtag|ga\(/i',
            'Google Tag Manager' => '/googletagmanager/i',
            'Facebook Pixel' => '/facebook\.net\/tr|fbq\(/i',
            'Hotjar' => '/hotjar/i'
        ];

        foreach ($analytics_patterns as $service => $pattern) {
            if (preg_match($pattern, $content)) {
                return $service;
            }
        }

        return 'None detected';
    }
}

// Handle the request
if ($_SERVER['REQUEST_METHOD'] === 'POST') {
    $input = json_decode(file_get_contents('php://input'), true);
    
    if (!isset($input['url'])) {
        echo json_encode(['success' => false, 'error' => 'URL parameter required']);
        exit;
    }

    $scanner = new WebsiteScanner($input['url']);
    $result = $scanner->scan();
    
    echo json_encode($result);
} else {
    echo json_encode(['success' => false, 'error' => 'Only POST method allowed']);
}
?>

File: scanner.py
#!/usr/bin/env python3
import json
import sys
import urllib.request
import urllib.parse
import socket
import ssl
import re
from datetime import datetime
import subprocess
import os

class WebsiteScanner:
    def __init__(self, url):
        self.url = url
        self.domain = urllib.parse.urlparse(url).netloc
        self.results = {}

    def scan(self):
        """Main scanning method"""
        try:
            self.results['basic_info'] = self.get_basic_info()
            self.results['server_info'] = self.get_server_info()
            self.results['domain_info'] = self.get_domain_info()
            self.results['security_info'] = self.get_security_info()
            self.results['performance_info'] = self.get_performance_info()
            self.results['technology_info'] = self.get_technology_info()
            
            return {
                'success': True,
                'data': self.results,
                'scan_time': datetime.now().isoformat()
            }
        except Exception as e:
            return {
                'success': False,
                'error': str(e)
            }

    def get_basic_info(self):
        """Get basic website information"""
        parsed_url = urllib.parse.urlparse(self.url)
        
        try:
            response = urllib.request.urlopen(self.url, timeout=10)
            status_code = response.getcode()
        except Exception:
            status_code = 'Unknown'

        return {
            'url': self.url,
            'domain': self.domain,
            'status_code': status_code,
            'protocol': parsed_url.scheme,
            'port': parsed_url.port or (443 if parsed_url.scheme == 'https' else 80)
        }

    def get_server_info(self):
        """Get server information"""
        try:
            ip_address = socket.gethostbyname(self.domain)
        except Exception:
            ip_address = 'Unknown'

        try:
            response = urllib.request.urlopen(self.url, timeout=10)
            server = response.headers.get('Server', 'Unknown')
        except Exception:
            server = 'Unknown'

        # Get location info (simplified)
        location_info = self.get_location_info(ip_address)

        return {
            'ip_address': ip_address,
            'server': server,
            'location': location_info.get('country', 'Unknown'),
            'city': location_info.get('city', 'Unknown'),
            'isp': location_info.get('isp', 'Unknown')
        }

    def get_domain_info(self):
        """Get domain information"""
        # Simplified domain info
        # In real implementation, you'd use whois libraries
        return {
            'registrar': 'Unknown',
            'creation_date': 'Unknown',
            'expiry_date': 'Unknown',
            'name_servers': self.get_name_servers(),
            'dnssec': 'Unknown'
        }

    def get_security_info(self):
        """Get security information"""
        ssl_info = self.get_ssl_info()
        security_headers = self.get_security_headers()

        return {
            'ssl_certificate': ssl_info,
            'security_headers': security_headers,
            'hsts_enabled': 'Strict-Transport-Security' in security_headers,
            'content_security_policy': 'Content-Security-Policy' in security_headers
        }

    def get_performance_info(self):
        """Get performance information"""
        import time
        
        start_time = time.time()
        try:
            response = urllib.request.urlopen(self.url, timeout=10)
            response_time = (time.time() - start_time) * 1000
            content_length = response.headers.get('Content-Length', 'Unknown')
        except Exception:
            response_time = 'Unknown'
            content_length = 'Unknown'

        return {
            'response_time': f"{response_time:.2f}ms" if response_time != 'Unknown' else 'Unknown',
            'compression': self.check_compression(),
            'caching': self.check_caching(),
            'page_size': f"{int(content_length)/1024:.2f} KB" if content_length != 'Unknown' and content_length.isdigit() else 'Unknown'
        }

    def get_technology_info(self):
        """Get technology stack information"""
        try:
            response = urllib.request.urlopen(self.url, timeout=10)
            content = response.read().decode('utf-8', errors='ignore')
            headers = dict(response.headers)
        except Exception:
            content = ''
            headers = {}

        return {
            'web_server': self.detect_web_server(headers),
            'programming_language': self.detect_programming_language(headers, content),
            'cms': self.detect_cms(content),
            'javascript_frameworks': self.detect_js_frameworks(content),
            'analytics': self.detect_analytics(content)
        }

    def get_location_info(self, ip):
        """Get geolocation information for IP"""
        # Simplified location detection
        # In real implementation, use services like ipapi.co
        return {
            'country': 'Unknown',
            'city': 'Unknown',
            'isp': 'Unknown'
        }

    def get_name_servers(self):
        """Get domain name servers"""
        try:
            import subprocess
            result = subprocess.run(['nslookup', '-type=ns', self.domain], 
                                  capture_output=True, text=True, timeout=10)
            # Parse nslookup output for nameservers
            nameservers = []
            for line in result.stdout.split('\n'):
                if 'nameserver' in line.lower():
                    ns = line.split()[-1].rstrip('.')
                    nameservers.append(ns)
            return ', '.join(nameservers) if nameservers else 'Unknown'
        except Exception:
            return 'Unknown'

    def get_ssl_info(self):
        """Get SSL certificate information"""
        if not self.url.startswith('https'):
            return {'enabled': False, 'message': 'SSL not enabled'}

        try:
            context = ssl.create_default_context()
            with socket.create_connection((self.domain, 443), timeout=10) as sock:
                with context.wrap_socket(sock, server_hostname=self.domain) as ssock:
                    cert = ssock.getpeercert()
                    
                    return {
                        'enabled': True,
                        'issuer': cert.get('issuer', [{}])[0].get('commonName', 'Unknown'),
                        'valid_from': cert.get('notBefore', 'Unknown'),
                        'valid_to': cert.get('notAfter', 'Unknown'),
                        'subject': cert.get('subject', [{}])[0].get('commonName', 'Unknown')
                    }
        except Exception as e:
            return {'enabled': False, 'message': f'SSL error: {str(e)}'}

    def get_security_headers(self):
        """Get security headers"""
        try:
            response = urllib.request.urlopen(self.url, timeout=10)
            headers = dict(response.headers)
            
            security_headers = {}
            security_header_names = [
                'Strict-Transport-Security',
                'Content-Security-Policy',
                'X-Frame-Options',
                'X-Content-Type-Options',
                'X-XSS-Protection',
                'Referrer-Policy'
            ]
            
            for header_name in security_header_names:
                if header_name in headers:
                    security_headers[header_name] = headers[header_name]
                    
            return security_headers
        except Exception:
            return {}

    def check_compression(self):
        """Check if compression is enabled"""
        try:
            response = urllib.request.urlopen(self.url, timeout=10)
            return 'Enabled' if 'Content-Encoding' in response.headers else 'Disabled'
        except Exception:
            return 'Unknown'

    def check_caching(self):
        """Check caching headers"""
        try:
            response = urllib.request.urlopen(self.url, timeout=10)
            cache_headers = ['Cache-Control', 'Expires', 'ETag', 'Last-Modified']
            
            for header in cache_headers:
                if header in response.headers:
                    return 'Enabled'
            return 'Disabled'
        except Exception:
            return 'Unknown'

    def detect_web_server(self, headers):
        """Detect web server"""
        server = headers.get('Server', 'Unknown')
        return server.split('/')[0] if server != 'Unknown' else 'Unknown'

    def detect_programming_language(self, headers, content):
        """Detect programming language"""
        # Check headers
        powered_by = headers.get('X-Powered-By', '')
        if 'PHP' in powered_by:
            return 'PHP'
        elif 'ASP.NET' in powered_by:
            return 'ASP.NET'
        
        # Check content
        if re.search(r'\.php["\']', content):
            return 'PHP'
        elif re.search(r'\.aspx["\']', content):
            return 'ASP.NET'
        elif re.search(r'\.jsp["\']', content):
            return 'Java'
        elif re.search(r'\.py["\']', content):
            return 'Python'
        
        return 'Unknown'

    def detect_cms(self, content):
        """Detect Content Management System"""
        cms_patterns = {
            'WordPress': r'wp-content|wp-includes|wordpress',
            'Drupal': r'drupal|sites/default',
            'Joomla': r'joomla|option=com_',
            'Magento': r'magento|mage/js',
            'Shopify': r'shopify|cdn\.shopify'
        }
        
        for cms, pattern in cms_patterns.items():
            if re.search(pattern, content, re.IGNORECASE):
                return cms
                
        return 'Custom/Unknown'

    def detect_js_frameworks(self, content):
        """Detect JavaScript frameworks"""
        frameworks = []
        framework_patterns = {
            'React': r'react|reactjs',
            'Vue.js': r'vue\.js|vuejs',
            'Angular': r'angular|ng-',
            'jQuery': r'jquery'
        }
        
        for framework, pattern in framework_patterns.items():
            if re.search(pattern, content, re.IGNORECASE):
                frameworks.append(framework)
                
        return ', '.join(frameworks) if frameworks else 'Unknown'

    def detect_analytics(self, content):
        """Detect analytics services"""
        analytics_patterns = {
            'Google Analytics': r'google-analytics|gtag|ga\(',
            'Google Tag Manager': r'googletagmanager',
            'Facebook Pixel': r'facebook\.net/tr|fbq\(',
            'Hotjar': r'hotjar'
        }
        
        for service, pattern in analytics_patterns.items():
            if re.search(pattern, content, re.IGNORECASE):
                return service
                
        return 'None detected'

def main():
    """Main function for command line usage"""
    if len(sys.argv) != 2:
        print("Usage: python scanner.py <URL>")
        sys.exit(1)
    
    url = sys.argv[1]
    scanner = WebsiteScanner(url)
    result = scanner.scan()
    
    print(json.dumps(result, indent=2))

if __name__ == "__main__":
    main()